README.md
# Zway-ThermostatControl

Control multiple radiator valves or thermostats using a virtual thermostat
using advanced schedules based on presence modes, time and day of week. The
module will create a virtual thermostat device that allows manual overriding
of the calculated setpoints. It also creates a binary switch that allows
to entirely disable the thermostat (eg. for summer or for manual overrides)

Multiple radiator valves or thermostats may be grouped in a zone. These zones
may have additional schedules that either override (absolute setpoints) or 
augment the global schedules (relative setpoints)

# Configuration

This module requires the Presence module 
(see https://github.com/maros/Zway-Presence), or any other module that behaves
the same to be installed first. If you don't want to use the given presence 
module, just must create a virtual switchBinary device, using 'Presence' as a 
probeTitle and use 'metrics:mode' to store the presence mode. The module 
should also emit events on switching modes (see Events)

## unitTemperature

Handle temperatures in Celsius or Fahrenheit

## defaultTemperature

Fallback default temperature. If no specific schedule is found, this 
temperature will be used.

## globalLimit.maxTemperature, globalLimit.minTemperature

Allows to specify optional upper and lower bounds for the virtual thermostat
and each controlled thermostat.

## globalSchedules

A list of global schedules. A schedule may apply to one or multiple presence 
modes, days of the week and have a starting and end time. Every schedule has 
a setpoint. If schedule rules overlap, only the first matching schedule will 
be processed. The order of schedules therefore matters.

## globalSchedules.presenceMode

List of presence modes (see https://github.com/maros/Zway-Presence) that this
schedule applies to. ie. allows to define lower temperatures for vacations and
away modes. Supported presence modes are

* Home: At home during daytime
* Night: At Home during nighttime / Sleeping
* Away: Away - both day and night)
* Vacation: Prolonged absence - both day and night

## globalSchedules.dayofweek

Days of the week that this schedule applies to. ie. allows to to specify
higher temperatures for the weekend.

## globalSchedules.timeFrom, globalSchedules.timeTo

Allows to specify a time (in HH:MM) when the schedule should be active.

## globalSchedules.mode, globalSchedules.setpoint

Based on the mode, a setpoint can be either absolute (eg. 19°C) or relative 
(eg. -1°C). Relative setpoints augment the defaultTemperature (and global 
setpoint for zone schedules).

eg. if the global setpoint is 20°C, and the relative zone setpoint is -2°C, 
then all radiator valves in the given zone will be set to 18°C.

In order to allow for manual thermostat overrides it is advisable to
use absolute setpoints for global schedules, and relative setpoints for
zone schedules.

## zones

Specify multiple temperature zones. Each zone may have additional schedules 
that either override (absolute setpoints), or augment (relative setpoints) 
the global schedules.

## zones.limit.maxTemperature, zones.limit.minTemperature

Every zone may have its own upper and lower bounds. If no values are set,
globalLimit.maxTemperature and globalLimit.minTemperature will be used.

## devices

List of thermostats that are managed in this zone.

## zones.schedules

Multiple schedules that override the global schedules. See globalSchedules
for documentation.

# Virtual Devices

This module creates a virtual thermostat which lets you manually override the 
global setpoint. The manual override will end once the next global schedule 
change occurs.

A binary switch will also be created, allowing for disabling automated
thermostat setpoints. (eg. for summer or for manual overrides)

# Events

No events are emitted.

The module listens to the following events, which are usually emitted by
the Presence module ( https://github.com/maros/Zway-Presence ), but may 
originate from any other module.

* presence.home
* presence.away
* presence.vacation
* presence.night

# Installation

Install the BaseModule from https://github.com/maros/Zway-BaseModule first

The prefered way of installing this module is via the "Zwave.me App Store"
available in 2.2.0 and higher. For stable module releases no access token is 
required. If you want to test the latest pre-releases use 'k1_beta' as 
app store access token.

For developers and users of older Zway versions installation via git is 
recommended.

```shell
cd /opt/z-way-server/automation/userModules
git clone https://github.com/maros/Zway-ThermostatControl.git ThermostatControl --branch latest
```

To update or install a specific version
```shell
cd /opt/z-way-server/automation/userModules/ThermostatControl
git fetch --tags
# For latest released version
git checkout tags/latest
# For a specific version
git checkout tags/1.02
# For development version
git checkout -b master --track origin/master
```

# License

Thermometer icon by Dianne Kathleen Navarro from the Noun Project

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or any 
later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.


CHANGES
1.07 2016 Oct 06, Thu
    * Fix day of week scheduling
    * Correctly set min/max temp and step for thermostat
    * Update icon
    * Improve texts
    * Set defaults

1.06 2016 Apr 21, Thu
    * Use dateFormat/mask/placeholder and correct type for time fields
    * Calculate setpoint on controller on
    * Set probeType
    * Update installation howto
    * Add doc link
    * Improve en translation
    * German translation

1.05 2016 Jan 18, Mon
    * Declare cron dependency
    * Rename title/description key to make module work in appstore

1.04 2016 Jan 02, Sat
    * Cleanup min/max temp
    * Use cron instead of timeouts to handle time schedules
    * Remember calculated setpoint. Do not re-calculate manual setpoints on init
    * Update documentation
    * Round to 0.5 degrees
    * Fix init
    * Fix module.json for 2.2.0
    * Use methods from BaseModule
    * Do not calculate global setpoint on manual change
    * Use BaseModule
    * cleanup jshint issues
    * Set probeType, Move variable elements to overlay instead of defaults
    * Improve parsing and processing of time settings
    * Unify variable names

1.03 2015 Nov 15, Sun
    * Fix presence device detection
    * Update documentation
    * Cleanups

1.02 2015 Nov 03, Tue
    * Fix schedule calculation
    * Update documentation
    * Add on/off vDev

1.01 2015 Nov 01, Sun
   * Initial release


index.js
/*** ThermostatControl Z-Way HA module *******************************************

Version: 1.07
(c) Maroš Kollár, 2015
-----------------------------------------------------------------------------
Author: Maroš Kollár <maros@k-1.com>
Description:
    Control multiple radiator valves (or thermostats) using a virtual thermostat.

******************************************************************************/

function ThermostatControl (id, controller) {
    // Call superconstructor first (AutomationModule)
    ThermostatControl.super_.call(this, id, controller);

    this.minTemperature     = undefined;
    this.maxTemperature     = undefined;
    this.vDevThermostat     = undefined;
    this.vDevSwitch         = undefined;
    this.cronName           = undefined;
}

inherits(ThermostatControl, BaseModule);

_module = ThermostatControl;

// ----------------------------------------------------------------------------
// --- Module instance initialized
// ----------------------------------------------------------------------------

ThermostatControl.prototype.init = function (config) {
    ThermostatControl.super_.prototype.init.call(this, config);
    var self = this;

    self.minTemperature = parseFloat(config.globalLimit.minTemperature)
        || (config.unitTemperature === 'celsius' ? 10 : 50);
    self.maxTemperature = parseFloat(config.globalLimit.maxTemperature)
        || (config.unitTemperature === 'celsius' ? 30 : 85);

    // Create vdev thermostat
    self.vDevThermostat = self.controller.devices.create({
        deviceId: "ThermostatControl_Thermostat_" + self.id,
        defaults: {
            metrics: {
                calculatedLevel: config.defaultTemperature,
                level: config.defaultTemperature,
                icon: 'thermostat',
                title: self.langFile.m_title
            },
        },
        overlay: {
            metrics: {
                min: self.minTemperature,
                max: self.maxTemperature,
                step: self.config.unitTemperature === 'celsius' ? 0.5 : 1,
                scaleTitle: config.unitTemperature === "celsius" ? '°C' : '°F'
            },
            probeType: 'thermostat_set_point',
            deviceType: 'thermostat'
        },
        handler: function(command, args) {
            if (command === 'exact') {
                var level = self.checkLimit(parseFloat(args.level));
                self.log('Manually changing setpoint to '+level);
                self.vDevThermostat.set("metrics:level", level);
                self.calculateSetpoint('setpoint');
            }
        },
        moduleId: this.id
    });

    // Create vdev switch
    self.vDevSwitch = self.controller.devices.create({
        deviceId: "ThermostatControl_Switch_" + self.id,
        defaults: {
            metrics: {
                level: 'on',
                icon: 'thermostat',
                title: self.langFile.m_title
            },
        },
        overlay: {
            deviceType: 'switchBinary',
            probeType: 'thermostat_mode'
        },
        handler: function(command, args) {
            var oldLevel = this.get('metrics:level');
            if (command === 'on' || command === 'off') {
                this.set('metrics:level',command);
                if (oldLevel !== command && command === 'on') {
                    self.calculateSetpoint('setpoint');
                }
            }
        },
        moduleId: this.id
    });

    self.cronName       = 'ThermostatControl.'+self.id+'.cron';
    self.callbackEvent  = _.bind(self.calculateSetpoint,self);

    self.controller.on(self.cronName,self.callbackEvent);

    // Init presence change callbacks
    _.each(self.presenceModes,function(presenceMode) {
        self.controller.on("presence."+presenceMode, self.callbackEvent,"presence");
    });

    // Init cron times
    _.each(self.config.globalSchedules,function(schedule) {
        self.initSchedule(schedule,'global');
    });

    _.each(self.config.zones,function(zone,index) {
        _.each(zone.schedules,function(schedule) {
            self.initSchedule(schedule,'zone.'+index);
        });
    });

    setTimeout(self.callbackEvent,10000,'init');
};

ThermostatControl.prototype.stop = function() {
    var self = this;

    if (self.vDevThermostat) {
        self.controller.devices.remove(self.vDevThermostat.id);
        self.vDevThermostat = undefined;
    }

    if (self.vDevSwitch) {
        self.controller.devices.remove(self.vDevSwitch.id);
        self.vDevSwitch = undefined;
    }

    _.each(self.presenceModes,function(presenceMode) {
        self.controller.off("presence."+presenceMode, self.callbackEvent);
    });

    self.controller.off(self.cronName,self.callbackEvent);
    self.controller.emit("cron.removeTask",self.cronName);

    self.callbackEvent = undefined;

    ThermostatControl.super_.prototype.stop.call(this);
};

// ----------------------------------------------------------------------------
// --- Module methods
// ----------------------------------------------------------------------------

ThermostatControl.prototype.initSchedule = function(schedule,id) {
    var self = this;

    if (typeof(schedule.timeFrom) === 'undefined'
        || typeof(schedule.timeTo) === 'undefined')
        return;

    _.each(['timeFrom','timeTo'],function(timeString) {
        var date        = self.parseTime(schedule[timeString]);
        if (schedule.dayofweek.length === 0
            || schedule.dayofweek.length === 7) {
            self.controller.emit("cron.addTask",self.cronName, {
                minute:     date.getMinutes(),
                hour:       date.getHours(),
                weekDay:    null,
                day:        null,
                month:      null,
            },id);
        } else {
            _.each(schedule.dayofweek,function(dayofweek) {
                self.controller.emit("cron.addTask",self.cronName, {
                    minute:     date.getMinutes(),
                    hour:       date.getHours(),
                    weekDay:    parseInt(dayofweek,10),
                    day:        null,
                    month:      null,
                },id);
            });
        }
    });
};

ThermostatControl.prototype.calculateSetpoint = function(source) {
    var self = this;

    if (self.vDevSwitch.get('metrics:level') === 'off') {
        self.log('Skipping setpoint calculation');
        return;
    }

    source = source || 'unknown';
    source = source.toString();
    self.log('Calculating setpoints due to '+source);

    var fromZone        = source.match(/^zone\.[0-9]+$/);
    var dateNow         = new Date();
    var dayNow          = dateNow.getDay();
    var presenceNow     = self.getPresenceMode();
    var curSetpoint     = self.vDevThermostat.get('metrics:level');
    var calcSetpoint    = self.vDevThermostat.get('metrics:calculatedLevel');
    var globalSetpoint  = self.config.defaultTemperature;

    var evalSchedule    = function(schedule) {
        // Check presence mode
        if (typeof(schedule.presenceMode) === 'object'
            && schedule.presenceMode.length > 0
            && _.indexOf(schedule.presenceMode, presenceNow) === -1) {
            return false;
        }

        // Check day of week if set
        if (typeof(schedule.dayofweek) === 'object'
            && schedule.dayofweek.length > 0
            && _.indexOf(schedule.dayofweek, dayNow.toString()) === -1) {
            return false;
        }

        // Check from/to time
        return self.checkPeriod(schedule.timeFrom,schedule.timeTo);
    };

    // Find global schedules & set global setpoint
    if (source !== 'setpoint'
        && ! fromZone) {
        _.find(self.config.globalSchedules,function(schedule) {
            if (evalSchedule(schedule) === false) {
                self.log('No global match');
                return;
            }
            if (schedule.mode === 'absolute') {
                globalSetpoint = parseFloat(schedule.setpoint);
            } else if (schedule.mode === 'relative') {
                globalSetpoint = self.config.defaultTemperature + parseFloat(schedule.setpoint);
            }
            globalSetpoint = self.checkLimit(globalSetpoint);
            return true;
        });
        self.vDevThermostat.set('metrics:calculatedLevel',globalSetpoint);
        // Change setpoint
        if (curSetpoint !== globalSetpoint) {

            // Was changed manually - not going to change on restart
            if (source === 'init'
                && calcSetpoint !== curSetpoint) {
                self.log('Not changing manual global setpoint to '+globalSetpoint);
                globalSetpoint = curSetpoint;
            } else {
                self.log('Changing global setpoint to '+globalSetpoint);
                self.vDevThermostat.set('metrics:level',globalSetpoint);
            }
        }
    } else {
        globalSetpoint = curSetpoint;
    }

    // Process zones
    _.each(self.config.zones,function(zone,index) {
        var zoneSetpoint = globalSetpoint;

        if (source === 'zone.'+index
            || ! fromZone) {
            // Find zone schedules
            _.find(zone.schedules,function(schedule) {
                if (evalSchedule(schedule) === false) {
                    return;
                }
                if (schedule.mode === 'absolute') {
                    zoneSetpoint = parseFloat(schedule.setpoint);
                } else if (schedule.mode === 'relative') {
                    zoneSetpoint = globalSetpoint + parseFloat(schedule.setpoint);
                }
                return zoneSetpoint;
            });
            zoneSetpoint = self.checkLimit(zoneSetpoint,zone.limit);
            self.log('Changing zone '+index+' to '+zoneSetpoint);

            // Set devices
            self.processDeviceList(zone.devices,function(deviceObject) {
                self.log('Setting '+deviceObject.get('metrics:title')+' to '+zoneSetpoint);
                deviceObject.set('metrics:calculatedLevel',zoneSetpoint);
                deviceObject.performCommand('exact',{ level: zoneSetpoint });
            });
        }
    });

};

ThermostatControl.prototype.checkLimit = function(level,limit) {
    var self = this;
    level = parseFloat(level);

    // TODO fallback limits?
    limit   = limit || {};
    var max = limit.maxTemperature || self.maxTemperature;
    var min = limit.minTemperature || self.minTemperature;

    if (typeof(max) === 'number'
        && level > max) {
        level = max;
    } else if (typeof(min) === 'number'
        && level < min) {
        level = min;
    }

    return Math.round(level*2)/2;
};

